20180634_최준혁_응용미시계량분석_HW1

# Verbal Questions
Q. 회사에서 건강보험을 제공한 여부가 임금에 영향을 미치는 지 확인하는 회귀분석에서 X는 개인과 회사의 특성이 담긴 독립변수 벡터일 때 도구변수로 firm size와 union status는 어떠한지 실증적인 관점에서 validity를 확인해라 

Ans : 먼저 firm size가 커질 수록 규모가 큰 회사는 직원들에게 insurance를 제공하는 경향이 있고 union status가 높을 수록 회사에는 중요 인력이라고 판단해 insurance를 제공하는 경향이 있을 것이다 따라서 두 도구변수와 insurance는 relevance가 있다. 하지만 도구변수는 종속변수인 임금에 대한 직접적인 영향이 없어야하는데 회사 내 개인 직급과 회사의 크기는 임금에 직접적인 영향을 주므로 Exclusion restriction을 위배한다. 개인이 직급이 높을 수록 높은 임금을 받는 경향이 뚜렷하고 또한 회사의 크기(직원 수)역시 회사의 규모가 큰 대기업일 수록 높은 임금을 제공하는 경향이 있으므로 종속변수에 직접적인 영향을 준다. 따라서 두 변수 모두 도구 변수로 타당하지 않다 따라서 회사 내 개인직급과 회사의 크기는 실증적인 관점에서 타당성이 떨어진다. 

# Computational Questions

## (Wooldridge (2019), Problem 3.C6) Use the data set in WAGE2 for this problem.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline

In [2]:
plt.rc('font', family='Malgun Gothic')
plt.rcParams['axes.unicode_minus'] = False

In [3]:
data = pd.read_csv('wage2.csv')
df_wage = data.copy()
df_wage.head()

wage  hours   IQ  KWW  educ  exper  tenure  age  married  black  south  \
0   769     40   93   35    12     11       2   31        1      0      0   
1   808     50  119   41    18     11      16   37        1      0      0   
2   825     40  108   46    14     11       9   33        1      0      0   
3   650     40   96   32    12     13       7   32        1      0      0   
4   562     40   74   27    11     14       5   34        1      0      0   

   urban  sibs  brthord  meduc  feduc     lwage  
0      1     1      2.0    8.0    8.0  6.645091  
1      1     1      NaN   14.0   14.0  6.694562  
2      1     1      2.0   14.0   14.0  6.715383  
3      1     4      3.0   12.0   12.0  6.476973  
4      1    10      6.0    6.0   11.0  6.331502

### (a) Run a simple regression of IQ on educ to obtain the slope coefficient, say, ˜δ1.
- ans : ˜δ1 = 3.5338

In [4]:
#회귀분석 실시 - statemodels 사용
from statsmodels.formula.api import ols
import statsmodels.api as sm

model = ols('IQ ~ educ', df_wage)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     IQ   R-squared:                       0.266
Model:                            OLS   Adj. R-squared:                  0.265
Method:                 Least Squares   F-statistic:                     338.0
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           1.16e-64
Time:                        21:49:40   Log-Likelihood:                -3717.0
No. Observations:                 935   AIC:                             7438.
Df Residuals:                     933   BIC:                             7448.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     53.6872      2.623     20.468      0.000      48.540      58.835
educ           3.5338      0.192     18.385      0.000       3.157       3.911
==============================================================================
Omnibus:                       30.954   Durbin-Watson:                   1.779
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               35.266
Skew:                          -0.398   Prob(JB):                     2.20e-08
Kurtosis:                       3.522   Cond. No.                         85.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### (b) Run the simple regression of log(wage) on educ, and obtain the slope coefficient, β˜1.
- ans : β˜1 = 0.0598

In [5]:
model = ols('lwage ~ educ', df_wage)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  lwage   R-squared:                       0.097
Model:                            OLS   Adj. R-squared:                  0.096
Method:                 Least Squares   F-statistic:                     100.7
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           1.42e-22
Time:                        21:49:40   Log-Likelihood:                -469.72
No. Observations:                 935   AIC:                             943.4
Df Residuals:                     933   BIC:                             953.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      5.9731      0.081     73.403      0.000       5.813       6.133
educ           0.0598      0.006     10.035      0.000       0.048       0.072
==============================================================================
Omnibus:                       31.006   Durbin-Watson:                   1.779
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               37.262
Skew:                          -0.375   Prob(JB):                     8.10e-09
Kurtosis:                       3.627   Cond. No.                         85.3
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### (c) Run the multiple regression of log(wage) on educ and IQ, and obtain the slope coefficients, βˆ1 and βˆ2, respectively.
- ans : βˆ1 = 0.0391, βˆ2 = 0.0059

In [6]:
model = ols('lwage ~ educ + IQ', df_wage)
result = model.fit()
result.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  lwage   R-squared:                       0.130
Model:                            OLS   Adj. R-squared:                  0.128
Method:                 Least Squares   F-statistic:                     69.42
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           7.88e-29
Time:                        21:49:40   Log-Likelihood:                -452.72
No. Observations:                 935   AIC:                             911.4
Df Residuals:                     932   BIC:                             926.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      5.6583      0.096     58.793      0.000       5.469       5.847
educ           0.0391      0.007      5.721      0.000       0.026       0.053
IQ             0.0059      0.001      5.875      0.000       0.004       0.008
==============================================================================
Omnibus:                       35.757   Durbin-Watson:                   1.810
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               47.770
Skew:                          -0.376   Prob(JB):                     4.23e-11
Kurtosis:                       3.813   Cond. No.                         773.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### (d) Verify that β˜1 = βˆ1 + βˆ2 ·˜δ1.

β˜1 = 0.0598, βˆ1 = 0.0391, βˆ2 = 0.0059, , ˜δ1 = 3.5338 <br>
βˆ1 + βˆ2 ·˜δ1 = 0.0391 + 0.0059 x 3.5338 <br>
= 0.0599 = β˜1

## (Wooldridge (2019), Problem 15.C2) The data in FERTIL2 include, for women in Botswana during 1988, information on number of children, years of education, age, and religious and economic status variables.

In [7]:
data = pd.read_csv('fertil2.csv')
df_f = data.copy()
df_f.head()

mnthborn  yearborn  age  electric  radio   tv  bicycle  educ  ceb  \
0         5        64   24       1.0    1.0  1.0      1.0    12    0   
1         1        56   32       1.0    1.0  1.0      1.0    13    3   
2         7        58   30       1.0    0.0  0.0      0.0     5    1   
3        11        45   42       1.0    0.0  1.0      0.0     4    3   
4         5        45   43       1.0    1.0  1.0      1.0    11    2   

   agefbrth  ...  heduc  agesq  urban  urb_educ  spirit  protest  catholic  \
0       NaN  ...    NaN    576      1        12       0        0         0   
1      25.0  ...   12.0   1024      1        13       0        0         0   
2      27.0  ...    7.0    900      1         5       1        0         0   
3      17.0  ...   11.0   1764      1         4       0        0         0   
4      24.0  ...   14.0   1849      1        11       0        1         0   

   frsthalf  educ0  evermarr  
0         1      0         0  
1         1      0         1  
2         0      0         1  
3         0      0         1  
4         1      0         1  

[5 rows x 27 columns]

In [8]:
df_f = df_f[['children','educ', 'age' , 'agesq','frsthalf','electric', 'tv', 'bicycle']]
df_f.dropna(inplace = True)
df_f = sm.add_constant(df_f, has_constant="add")

### (a) interpret the estimates. In particular, holding age fixed, what is the estimated effect of another year of education on fertility?
- ans : 특히! 다른 독립변수들(age, age^2)가 고정되어 있을때, 여성의 교육년수가 1년 증가하면 출산하는 아이가 0.0904명 감소한다
<br> 전체 F값과 개별 변수의 t값의 p-value가 매우 낮으므로 유의하며 변수들이 fertility 총변동의 56%를 설명하고 있다
<br> 다른 독립변수들이 고정되어 있을때, 여성의 나이가 1년 증가하면 출산하는 아이가 0.33명 증가한다.
<br> 다른 독립변수들이 고정되어 있을때, 여성의 agesq가 한단위 증가하면 출산하는 아이가 0.0026명 감소한다.

In [9]:
from linearmodels.iv import IV2SLS

OLS = IV2SLS(df_f.children, df_f[["const","educ", "age","agesq"]], None, None).fit(cov_type="unadjusted")
print(OLS)

                            OLS Estimation Summary                            
Dep. Variable:               children   R-squared:                      0.5684
Estimator:                        OLS   Adj. R-squared:                 0.5681
No. Observations:                4356   F-statistic:                    5736.7
Date:                Mon, Oct 10 2022   P-value (F-stat)                0.0000
Time:                        21:49:42   Distribution:                  chi2(3)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const         -4.1417     0.2406    -17.213     0.0000     -4.6133     -3.6701
educ          -0.0904     0.0059    -15.252     0.00

### (b) The variable frsthalf is a dummy variable equal to one if the woman was born during the first six months of the year. Assuming that frsthalf is uncorrelated with the error term from part (a), how would you validate the relevance condition for this variable to be an IV for educ?
- ans : 에러텀과 frsthalf가 연관되어있지 않다고 가정하였으므로 frsthalf이 educ과 cov가 0이 아님을 밝히면 도구변수로서 타당함을 밝힐 수 있다. 따라서 educ에 대해 frsthalf로 regression한 결과는 다음과 같으며 따라서 frsthalf를 educ의 도구변수로 두기에 타당하다.

In [10]:
stage1_model = ols('educ ~ frsthalf', df_f).fit()
stage1_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   educ   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.014
Method:                 Least Squares   F-statistic:                     62.86
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           2.79e-15
Time:                        21:49:42   Log-Likelihood:                -12107.
No. Observations:                4356   AIC:                         2.422e+04
Df Residuals:                    4354   BIC:                         2.423e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.3623      0.087     73.000      0.000       6.191       6.533
frsthalf      -0.9398      0.119     -7.929      0.000      -1.172      -0.707
==============================================================================
Omnibus:                       72.665   Durbin-Watson:                   1.397
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               41.623
Skew:                          -0.026   Prob(JB):                     9.16e-10
Kurtosis:                       2.524   Cond. No.                         2.72
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
reduced_model = ols('children ~ frsthalf', df_f).fit()
reduced_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               children   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     15.15
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           0.000101
Time:                        21:49:42   Log-Likelihood:                -9650.8
No. Observations:                4356   AIC:                         1.931e+04
Df Residuals:                    4354   BIC:                         1.932e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.1264      0.050     42.877      0.000       2.029       2.224
frsthalf       0.2625      0.067      3.892      0.000       0.130       0.395
==============================================================================
Omnibus:                      623.649   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              922.028
Skew:                           1.065   Prob(JB):                    6.08e-201
Kurtosis:                       3.739   Cond. No.                         2.72
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [12]:
x_hat = stage1_model.predict()
df_f['x_hat'] = x_hat
stage2_model = ols('children ~ x_hat', df_f).fit()
stage2_model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:               children   R-squared:                       0.003
Model:                            OLS   Adj. R-squared:                  0.003
Method:                 Least Squares   F-statistic:                     15.15
Date:                Mon, 10 Oct 2022   Prob (F-statistic):           0.000101
Time:                        21:49:42   Log-Likelihood:                -9650.8
No. Observations:                4356   AIC:                         1.931e+04
Df Residuals:                    4354   BIC:                         1.932e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      3.9037      0.421      9.261      0.000       3.077       4.730
x_hat         -0.2793      0.072     -3.892      0.000      -0.420      -0.139
==============================================================================
Omnibus:                      623.649   Durbin-Watson:                   1.976
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              922.028
Skew:                           1.065   Prob(JB):                    6.08e-201
Kurtosis:                       3.739   Cond. No.                         75.8
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### (c) Estimate the model from part (a) by using frsthalf as an IV for educ. Compare the estimated effect of education with the OLS estimate from part (a).
- ans : OLS의 educ계수 = -0.0904, IV estimation의 educ계수 = -0.2793

In [13]:
SLS = IV2SLS(df_f.children, df_f[["const"]], df_f.educ, df_f.frsthalf).fit(cov_type="unadjusted")
print(SLS)

                          IV-2SLS Estimation Summary                          
Dep. Variable:               children   R-squared:                      0.1214
Estimator:                    IV-2SLS   Adj. R-squared:                 0.1212
No. Observations:                4356   F-statistic:                    17.191
Date:                Mon, Oct 10 2022   P-value (F-stat)                0.0000
Time:                        21:49:42   Distribution:                  chi2(1)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const          3.9037     0.3957     9.8659     0.0000      3.1282      4.6792
educ          -0.2793     0.0674    -4.1462     0.00

In [14]:
from linearmodels.iv import compare

print(compare({"OLS": OLS, "2SLS":SLS}, precision = 'pvalues'))

                  Model Comparison                 
                                OLS            2SLS
---------------------------------------------------
Dep. Variable              children        children
Estimator                       OLS         IV-2SLS
No. Observations               4356            4356
Cov. Est.                unadjusted      unadjusted
R-squared                    0.5684          0.1214
Adj. R-squared               0.5681          0.1212
F-statistic                  5736.7          17.191
P-value (F-stat)             0.0000       3.381e-05
==================     ============   =============
const                       -4.1417          3.9037
                           (0.0000)        (0.0000)
educ                        -0.0904         -0.2793
                           (0.0000)     (3.381e-05)
age                          0.3327                
                           (0.0000)                
agesq                       -0.0026                
            

### (d) Add the binary variables electric, tv, and bicycle to the model and assume these are exogenous. Estimate the equation by OLS and 2SLS and compare the estimated coefficients on educ.
- ans : OLS의 educ계수 = -0.0767, IV estimation의 educ계수 = -0.2793

In [15]:
OLS = IV2SLS(df_f.children, df_f[["const","educ", "age","agesq",'electric','tv','bicycle']], None, None).fit(cov_type="unadjusted")
print(OLS)

                            OLS Estimation Summary                            
Dep. Variable:               children   R-squared:                      0.5761
Estimator:                        OLS   Adj. R-squared:                 0.5755
No. Observations:                4356   F-statistic:                    5919.0
Date:                Mon, Oct 10 2022   P-value (F-stat)                0.0000
Time:                        21:49:42   Distribution:                  chi2(6)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const         -4.3898     0.2401    -18.281     0.0000     -4.8604     -3.9191
educ          -0.0767     0.0063    -12.085     0.00

In [16]:
SLS = IV2SLS(df_f.children, df_f[["const"]], df_f.educ, df_f.frsthalf).fit(cov_type="unadjusted")
print(SLS)

                          IV-2SLS Estimation Summary                          
Dep. Variable:               children   R-squared:                      0.1214
Estimator:                    IV-2SLS   Adj. R-squared:                 0.1212
No. Observations:                4356   F-statistic:                    17.191
Date:                Mon, Oct 10 2022   P-value (F-stat)                0.0000
Time:                        21:49:42   Distribution:                  chi2(1)
Cov. Estimator:            unadjusted                                         
                                                                              
                             Parameter Estimates                              
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
const          3.9037     0.3957     9.8659     0.0000      3.1282      4.6792
educ          -0.2793     0.0674    -4.1462     0.00

In [17]:
print(compare({"OLS": OLS, "2SLS":SLS}, precision = 'pvalues'))

                  Model Comparison                  
                                 OLS            2SLS
----------------------------------------------------
Dep. Variable               children        children
Estimator                        OLS         IV-2SLS
No. Observations                4356            4356
Cov. Est.                 unadjusted      unadjusted
R-squared                     0.5761          0.1214
Adj. R-squared                0.5755          0.1212
F-statistic                   5919.0          17.191
P-value (F-stat)              0.0000       3.381e-05
==================     =============   =============
const                        -4.3898          3.9037
                            (0.0000)        (0.0000)
educ                         -0.0767         -0.2793
                            (0.0000)     (3.381e-05)
age                           0.3402                
                            (0.0000)                
agesq                        -0.0027          